tools/generators/xcschemes/src/Generator/CreateCustomSchemeInfos.swift

```diff
@@ -0,0 +1,619 @@
+
+        var environmentVariables: [EnvironmentVariable] = []
+        for _ in (0..<count) {
+            let key = try consumeArg(
+                String.self,
+                in: url,
+                file: file,
+                line: line
+            ).nullsToNewlines
+            let value = try consumeArg(
+                String.self,
+                in: url,
+                file: file,
+                line: line
+            ).nullsToNewlines
+            let enabled = try consumeArg(
+                Bool.self,
+                in: url,
+                file: file,
+                line: line
+            )
+
+            environmentVariables.append(
+                .init(key: key, value: value, enabled: enabled)
+            )
+        }
+
+        return environmentVariables
+    }
+
+    // MARK: - SchemeInfo.LaunchTarget
+
+    mutating func consumeArg(
+        _ type: SchemeInfo.LaunchTarget?.Type,
+        in url: URL,
+        allTargetIDs: inout Set<TargetID>,
+        context: @autoclosure () -> String,
+        commandLineArguments: inout [CommandLineArgument],
+        environmentVariables: inout [EnvironmentVariable],
+        extensionHostIDs: [TargetID: [TargetID]],
+        targetCommandLineArguments: [TargetID: [CommandLineArgument]],
+        targetEnvironmentVariables: [TargetID: [EnvironmentVariable]],
+        targetsByID: [TargetID: Target],
+        file: StaticString = #filePath,
+        line: UInt = #line
+    ) throws -> SchemeInfo.LaunchTarget? {
+        let id = try consumeArg(
+            TargetID?.self,
+            in: url,
+            file: file,
+            line: line
+        )
+        let extensionHostID = try consumeArg(
+            TargetID?.self,
+            in: url,
+            file: file,
+            line: line
+        )
+
+        guard let id else {
+            return nil
+        }
+
+        allTargetIDs.insert(id)
+
+        let target = try targetsByID.value(
+            for: id,
+            context: context()
+        )
+
+        guard !target.productType.needsExtensionHost || extensionHostID != nil
+        else {
+            throw UsageError(message: """
+\(context()) (\(id)) is an app extension and requires `extension_host` to be \
+set
+""")
+        }
+
+        let extensionHost = try extensionHostID.flatMap { extensionHostID in
+            guard extensionHostIDs[id, default: []]
+                .contains(where: { $0 == extensionHostID })
+            else {
+                throw UsageError(message: """
+\(context()) `extension_host` (\(extensionHostID)) does not host the extension \
+(\(id))
+""")
+            }
+            return try targetsByID.value(
+                for: extensionHostID,
+                context: "\(context()) extension host"
+            )
+        }
+
+        // Only set from-rule args and env if the custom scheme doesn't
+        // declare any
+        // TODO: Distinguish between `None` and `[]` in `xcshemes`?
```

Comment:
What would be the advantage or doing that instead of treating `None` and `[]` as the same "empty" input?

Reply:
Allows saying in the scheme that you don't want any arguments, regardless if the rule would provide some. Right now that's not possible, both `None` and `[]` would mean no custom ones, but you will get whatever the rule provides.

Reply:
I plan to have a special `inherit` value, which will also be the default. So if you set a falsey value, it won't inherit.

Reply:
SGTM, thx for clarifying!

Reply:
This is now in. `nil` means "inherit", and non-`nil` means that was explicitly specified by the user, so use those values. On the Starlark side `inherit` eventually becomes `nil` (because of a `-1` count), and `None`/`[]` both mean "don't inherit any args, I mean literally have no args for this action).